### PR TITLE
Further cardano testnet cleanup

### PR DESCRIPTION
**Pull request by Jimbo4350**
# Description

We now fail early in the tests when the `cardano-node` binary fails for some reason. I compared running the `ShutdownOnSigint` property and purposely made it fail with an invalid cardano-node flag (`--dataFFbase-path` instead of `--database-path`). Note that prior to these changes the test failure output did not report the failure. Previously you would need to look at the node's stderr output file in the directory the test was run in to figure out what went wrong.

This test now fails earlier (~46 seconds vs ~272 seconds).

After changes: 
```
...
   ["NodeExecutableError \"Invalid option `--dataFFbase-path'\\n\\nUsage: cardano-node (run | [--help])\\n\\n  Start node of the Cardano blockchain.\\n\"","NodeExecutableError \"Invalid option `--dataFFbase-path'\\n\\nUsage: cardano-node (run | [--help])\\n\\n  Start node of the Cardano blockchain.\\n\"","NodeExecutableError \"Invalid option `--dataFFbase-path'\\n\\nUsage: cardano-node (run | [--help])\\n\\n  Start node of the Cardano blockchain.\\n\""]
        
          This failure can be reproduced by running:
          > recheckAt (Seed 2006430009148179759 17883359916077017841) "1:" ShutdownOnSigint
        
      Use "--pattern '$NF ~ /ShutdownOnSigint/' --hedgehog-replay '1: Seed 2006430009148179759 17883359916077017841'" to reproduce from the command-line.

1 out of 1 tests failed (46.56s)
Test suite cardano-testnet-test: FAIL
Test suite logged to:
/home/jordan/Repos/Work/cardano-haskell/cardano-node/dist-newstyle/build/x86_64-linux/ghc-9.2.8/cardano-testnet-8.5.0/t/cardano-testnet-test/test/cardano-testnet-8.5.0-cardano-testnet-test.log
0 of 1 test suites (0 of 1 test cases) passed.
Error: cabal: Tests failed for test:cardano-testnet-test from
cardano-testnet-8.5.0.
```
Before changes:
```
...
      251 ┃   logs <- H.readFile (nodeStdout node)
          252 ┃   case mapMaybe parseMsg $ reverse $ lines logs of
          253 ┃     [] -> H.failMessage callStack "Could not find close DB message."
          254 ┃     (Left err):_ -> H.failMessage callStack err
          255 ┃     (Right _):_ -> pure ()
        
          This failure can be reproduced by running:
          > recheckAt (Seed 6798916397599032382 7647850035913318121) "1:" ShutdownOnSigint
        
      Use "--pattern '$NF ~ /ShutdownOnSigint/' --hedgehog-replay '1: Seed 6798916397599032382 7647850035913318121'" to reproduce from the command-line.

1 out of 1 tests failed (271.79s)
Test suite cardano-testnet-test: FAIL
Test suite logged to:
/home/jordan/Repos/Work/cardano-haskell/cardano-node/dist-newstyle/build/x86_64-linux/ghc-9.2.8/cardano-testnet-8.5.0/t/cardano-testnet-test/test/cardano-testnet-8.5.0-cardano-testnet-test.log
0 of 1 test suites (0 of 1 test cases) passed.
Error: cabal: Tests failed for test:cardano-testnet-test from
cardano-testnet-8.5.0.
```
# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - roundtrip tests
  - integration tests
  See [Runnings tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG node developers to do this
for you.

### CODE REVIEW

File: cardano-testnet/src/Testnet/Runtime.hs [hunk 1]
```diff
@@ -196,15 +202,15 @@ startNode tp@(TmpAbsolutePath _tempAbsPath) node port nodeCmd = GHC.withFrozenCa
...
-      nodeStderrFile = node <> ".stderr.log"
+  let nodeStdoutFile = logDir </> node <> ".stdout.log"
+      nodeStderrFile = logDir </> node <> ".stderr.log"
       sprocket = Sprocket tempBaseAbsPath (socketDir </> node)
 
   hNodeStdout <- handleIOExceptT FileRelatedFailure $ IO.openFile nodeStdoutFile IO.WriteMode
-  hNodeStderr <- handleIOExceptT FileRelatedFailure $ IO.openFile  nodeStderrFile IO.WriteMode
+  hNodeStderr <- handleIOExceptT FileRelatedFailure $ IO.openFile nodeStderrFile IO.ReadWriteMode
 
    -- H.diff (L.length (IO.sprocketArgumentName sprocket)) (<=) IO.maxSprocketArgumentNameLength
```
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Runtime.hs:215]: TODO: Need to add a failure for this

File: cardano-testnet/src/Testnet/Runtime.hs [hunk 2]
```diff
@@ -228,18 +233,46 @@ startNode tp@(TmpAbsolutePath _tempAbsPath) node port nodeCmd = GHC.withFrozenCa
...
+  if null stdErrContents
+  then return $ NodeRuntime node sprocket stdIn nodeStdoutFile nodeStderrFile hProcess
+  else left $ NodeExecutableError stdErrContents
 
-  -- The node process can fail on startup, e.g while
-  -- parsing the configuration yaml file. If we don't fail
-  -- when stderr is populated, we end up having to wait for the
-  -- timeout to expire before we can see the error.
   -- TODO: This is flakey. Hydra error:
   -- Unable to run finally: Failure Nothing "\9473\9473\9473 Exception (IOException) \9473\9473\9473\nlsof: readCreateProcess: posix_spawnp: illegal operation (Inappropriate ioctl for device)\n" Nothing
```
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Runtime.hs:255]: @newhoggy what was the intention of the call to `lsof`?

Comment (newhoggy) [reply cardano-testnet/src/Testnet/Runtime.hs:255]: It checks if the port is open.  We shouldn't continue the test if the port hasn't opened or we may fail to connect.

**Comment by Jimbo4350** [comment]
All checks passed but for some reason hydra did not update the github UI

**Pull request merged**